irohad/main/application.cpp

```diff
@@ -188,63 +193,79 @@ void Irohad::initStorage() {
       },
       block_converter,
       log_manager_);
-  auto storageResult = StorageImpl::create(block_store_dir_,
-                                           pg_conn_,
-                                           common_objects_factory_,
-                                           std::move(block_converter),
-                                           perm_converter,
-                                           std::move(block_storage_factory),
-                                           log_manager_->getChild("Storage"));
-  std::move(storageResult)
-      .match([&](auto &&v) { storage = std::move(v.value); },
-             [&](const auto &error) { log_->error("{}", error.error); });
-
-  log_->info("[Init] => storage ({})", logger::logBool(storage));
+
+  return StorageImpl::create(block_store_dir_,
+                             pg_conn_,
+                             common_objects_factory_,
+                             std::move(block_converter),
+                             perm_converter,
+                             std::move(block_storage_factory),
+                             log_manager_->getChild("Storage"))
+      .match(
+          [&](auto &&v) -> RunResult {
+            storage = std::move(v.value);
+            log_->info("[Init] => storage");
+            return {};
+          },
+          [&](auto &&error) -> RunResult { return std::move(error); });
 }
 
-bool Irohad::restoreWsv() {
+Irohad::RunResult Irohad::restoreWsv() {
   return wsv_restorer_->restoreWsv(*storage).match(
-      [](const auto &) { return true; },
-      [this](const auto &error) {
-        log_->error("{}", error.error);
-        return false;
-      });
+      [](const auto &value) -> RunResult {
+        if (not value.value) {
+          return iroha::expected::makeError<std::string>(
+              "Did not get ledger state after WSV restoration!");
+        }
+        auto &ledger_state = value.value.value();  // value
+        assert(ledger_state);
+        if (ledger_state->ledger_peers->empty()) {
+          return iroha::expected::makeError<std::string>(
+              "Have no peers in WSV after restoration!");
+        }
+        return {};
+      },
+      [](auto &&error) -> RunResult { return std::move(error); });
 }
 
-bool Irohad::resetPeers(
+Irohad::RunResult Irohad::resetPeers(
     const shared_model::interface::types::PeerList &alternative_peers) {
   storage->resetPeers();
+
+  RunResult reset_result = {};
   for (const auto &peer : alternative_peers) {
-    auto result = storage->insertPeer(*peer);
-    if (auto e = boost::get<expected::Error<std::string>>(&result)) {
-      log_->error("{}", e->error);
-      return false;
-    }
+    reset_result = reset_result | [&]() -> RunResult {
```

Comment:
This loop will perform useless iterations after the first error, please revert the changes.